### PR TITLE
Support `defer_build` for Pydantic dataclasses

### PR DESCRIPTION
**Pull request by Viicos**
Also return early if `defer_build` is set for Pydantic models.

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: pydantic/dataclasses.py [hunk 1]
```diff
@@ -321,11 +321,17 @@ def rebuild_dataclass(
...
         return _pydantic_dataclasses.complete_dataclass(
             cls,
             _config.ConfigWrapper(cls.__pydantic_config__, check=False),
             raise_errors=raise_errors,
             types_namespace=types_namespace,
+            # We could provide a different config instead (with `'defer_build'` set to `True`)
+            # but this doesn't play well in some cases as unlike Pydantic models, Pydantic dataclasses
+            # can have the config coming from the `config` argument or from the `__pydantic_config__`
+            # property. So we use an explicit argument:
+            _force_build=True,
```
Comment (Viicos) [inline pydantic/dataclasses.py:334]: For Pydantic models, we pass `{**config, "defer_build": False}` as the config, but for dataclasses this doesn't work well with this kind of scenario:

```python
@pydantic.dataclasses.dataclass(config=ConfigDict())
class Model:
    x: str
    __pydantic_config__ = ConfigDict(str_to_lower=True)
```

Hence the unfortunate new parameter.

File: tests/test_config.py [hunk 2]
```diff
@@ -889,7 +918,7 @@ class MyModel(BaseModel):
     assert exc_info.value.code == 'model-config-invalid-field-name'
 
 
-def test_dataclass_allowes_model_config_as_model_field():
+def test_dataclass_allows_model_config_as_model_field():
```
Comment (Viicos) [inline tests/test_config.py:921]: typo

File: pydantic/dataclasses.py (hunk 1 above)
Comment (sydney-runkle) [reply pydantic/dataclasses.py:334]: Hmm, I'd prefer we either disallow configs in both locations, or merge them with priority towards one.

That way, we could have one unified config for the model that we can patch with the `defer_build` when necessary.

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Looks great, left one comment re the `_force_build` argument - would like to avoid that if possible.

I've submitted https://github.com/pydantic/pydantic/pull/10329, which we can probably merge + rebase with to make this PR just a bit more simple.

File: pydantic/dataclasses.py (hunk 1 above)
Comment (sydney-runkle) [reply pydantic/dataclasses.py:334]: Should be able to move forward now that we've merged https://github.com/pydantic/pydantic/pull/10406

Comment (sydney-runkle) [reply pydantic/dataclasses.py]: Alright, I've updated this with the new changes!

**Comment by Viicos** [comment]
So the reason why the added `_force_build` argument was required is because currently ultimately, `rebuild_dataclass` calls `complete_dataclass`. Because the configuration can come from both the decorator or the `__pydantic_config__` attribute, `complete_dataclass` sets `cls.__pydantic_config__`:

https://github.com/pydantic/pydantic/blob/59a3de71c6592a0e088ce9c7c4c7dd1950a05b7b/pydantic/_internal/_dataclasses.py#L146

So that the following works as expected:

```python
@pydantic_dataclass(config={'strict': True}
class DC:
    a: int

assert DC.__pydantic_config__ == {'strict': True}
```

If in `rebuild_dataclass`, we pass a new config object, as done currently here:

https://github.com/pydantic/pydantic/blob/c2a325b720d70b692603f8150c161f2b9bd136a2/pydantic/dataclasses.py#L338-L345

The following will happen:

```python
@pydantic_dataclass
class DC:
    __pydantic_config__ = {'defer_build': True}  # or as a decorator parameter, doesn't matter

assert DC.__pydantic_config__ == {'defer_build': True}
rebuild_dataclass(DC)
assert DC.__pydantic_config__ == {'defer_build': True}  # AssertionError, `'defer_build'` is now set to `False`
```
And is inconsistent with the `BaseModel` behavior. Besides, I think the config dict on models and dataclasses shouldn't change after the class is defined, no matter if a rebuild was performed or not.

**Comment by sydney-runkle** [comment]
> And is inconsistent with the BaseModel behavior. Besides, I think the config dict on models and dataclasses shouldn't change after the class is defined, no matter if a rebuild was performed or not (and this is what's causing issues with TypeAdapter failing tests I think).

I see, makes sense. Let's revert back then to use the `_force_build` flag. Apologies for the back and forth on the diff.

**Review by sydney-runkle — APPROVED** [review]
Nice work!

**Pull request merged**